### PR TITLE
Vouchers part 1

### LINKED ISSUE #10857: [Vouchers] Errors when/from iterating between /summary and /cart page
## Description

_This is treated as one issue, as it seems to me (I might be wrong) that this is all coming from the same root cause. Also related to #10855_

After adding a voucher and while on the confirmation page, the user may opt for returning to either the /cart page (editing quantities, for ex.):

- If there are no taxes on the order/voucher -> voucher is removed from the order
- If there are taxes on the order/voucher, and `voucher amount` > `order.total` -> snail
- If there are taxes on the order/voucher, `voucher amount` <= `order.total` -> additional parcels will be created for the voucher, on each /summary page visit, till a snail is triggered

**Bugsnag**
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/6464d03f630d0f000895ebe9?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging&event_id=6464d90700bd31520e150000

## Expected Behavior

The user should be able to iterate between `/summary` and the `/cart` pages.

## Actual Behaviour

Iterating between the `/cart` and `/summary` pages triggers a snail.

## Steps to Reproduce

For all voucher issues: per-requisites are:
- As super admin - enable the vouchers feature toggle
- Having a hub with at least one voucher created
- The hub should have an open shop

An order/line item/customer with no taxes:

1. As a customer, on the shopfront, add items to the cart and proceed to the `/summary` step

image

2. Open the side menu and visit the /cart page. 

image

**no taxes on the order/voucher** 

3. Click Checkout button, to return to the /summary page -> notice the voucher was removed from the order.

image

**taxes on the order/voucher, and `voucher amount` > `order.total` -> snail**

3. Open the side menu and visit the /cart page. Click Checkout button -> :snail: 

image

## Animated Gif/Screenshot

image

## Workaround

- returning to the /payment step, and re-adding the voucher
- If error 500 is triggered, then the user is "stuck" -> only way out is emptying the cart on the `/cart` page

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Addresses a couple of things related to checkout processing and vouchers. This was previously combined with some other changes, but I've split them into two smaller chunks so it's easier to review. This is part one.

Partially fixes #10857, should be fully fixed by the next PR (#11003).

#### Review notes

Hopefully the commits are pretty well explained by their names but here's some extra notes:

- Removed an Angular loading indicator which in some cases was causing the screen to be blocked with a dark background and little loading animation in cases where it shouldn't be there, mostly on the order completed page. This was previously used by Angular but isn't needed any more.

- The customer details controller was doing a weird thing where it was handling applying taxes on admin adjustments, but that shouldn't be the responsibility of that controller, so its moved out to `#create_tax_charge!` where all other tax handling is done.

- Handling of the issue where _taxes on an order need to be recalculated if the order's address changes_ was being done solely by the customer details controller and that's now done at the model level, because it should apply to any context and not just that one controller.

- Taxes on enterprise fees don't get applied until the order is at least in `payment` state, eg it's passed `cart`, `address`, and `delivery` states. This is in line with how all tax handling works since split checkout was introduced.

- The way that re-calculating voucher amounts based on the order total was being handled was a bit off in certain cases where a voucher was already present, as it was using the order amount with discounts applied instead of the order amount before discounts. Eg: if the order total is 20 and a voucher for -5 is present and the order total changes to 22, when _recalculating_ the voucher amount it should be using the pre-discount total of 22 and not the discounted total of 22 - 5 (17) when making calculations.

#### What should we test?

- If an order has progressed to at least `payment` state then submitting a change to it's address should recalculate tax correctly if the change of address would change it's tax category. Note: which address gets used for tax purposes depends on an admin configuration setting.

- Some of the issues outlined in #10857 around errors when recalculating a voucher amount on an order should be resolved but maybe it's best to test that along with the next PR #11003 (and it's still behind a feature toggle here)?

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Nice. Looking great! :sparkles:

File: app/controllers/split_checkout_controller.rb [hunk 1]
```diff
@@ -204,6 +203,7 @@ def shipping_method_ship_address_not_required?
 
   def process_voucher
     if add_voucher
+      VoucherAdjustmentsService.calculate(@order)
```
Comment (rioug) [inline app/controllers/split_checkout_controller.rb:206]: As long as taxes don't change after this it should be fine. If I followed correctly taxes are recalculated if address changes, and also after the order move to the "payment" step, which all happen before we can reach this code, is that correct ?

**Review by rioug — APPROVED** [review]
Looks good :+1:

File: app/models/spree/order.rb [hunk 2]
```diff
@@ -171,6 +171,11 @@ def amount
       line_items.inject(0.0) { |sum, li| sum + li.amount }
     end
 
+    # Order total without any applied discounts from vouchers
+    def pre_discount_total
```
Comment (dacook) [inline app/models/spree/order.rb:175]: In case we have other types of "discount" in the system, I think a more specific name should be used here:

```suggestion
```

File: spec/system/consumer/split_checkout_spec.rb [hunk 3]
```diff
@@ -1114,11 +1114,8 @@
         let(:voucher) { create(:voucher, code: 'some_code', enterprise: distributor, amount: 6) }
 
         before do
-          # Add voucher to the order
```
Comment (dacook) [inline spec/system/consumer/split_checkout_spec.rb:1117]: These comments might seem superfluous to some, but as a first-time reader of the code, I find them a helpful narration :)

**Review by dacook — APPROVED** [review]
Great, looks good to me too! I have a suggestion for bonus points though ;)

File: spec/controllers/spree/admin/orders_controller_spec.rb [hunk 4]
```diff
@@ -57,7 +57,7 @@
 
         it "updates fees and taxes and redirects to order details page" do
           expect(order).to receive(:recreate_all_fees!)
-          expect(order).to receive(:create_tax_charge!)
+          expect(order).to receive(:create_tax_charge!).at_least :once
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply spec/controllers/spree/admin/orders_controller_spec.rb:60]: I puzzled over this a bit, but it looks like this happens in relation to what the test _setup_ is doing as opposed to what happens in the test itself. Basically before the test even starts (or when the order object is called and lazy-instantiated) the process of created a completed order (`:completed_order_with_totals`) is running through various bits of logic in  a sequence like adding line items and creating shipments and setting the shipping address, etc etc, and some of that is doing things that are triggering tax-related callbacks because the order total is changing or the tax address is changing.

File: app/models/spree/order.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply app/models/spree/order.rb:175]: Hmmm interesting! I don't really associate the word "discount" with the idea of percentages specifically, just with a reduction in price.

Yeah I guess at the moment the concepts "discount" and "voucher" are kind of synonymous in the codebase, vouchers apply a discount and the only kind of discount is from vouchers. But we have the `Voucher` class, and voucher objects that are saved on a _distributor_ and define a kind of "voucher rule", then we have voucher adjustments which are a different object altogether and are saved on an order.

As always, [naming things is hard](https://hilton.org.uk/blog/why-naming-things-is-hard).

I guess we could say that "voucher" is the _form_ and "discount" is the _function_? I think what the method is trying to represent is the **price** of the order prior to discounts that might modify that price, so conceptually I'm leaning more towards the function than the form in this case?

Hmmm so I'm not sure I completely agree, but then again I also don't completely disagree! :sweat_smile: 

Maybe the other devs have some opinions one way or the other on this naming conundrum..?

Comment (mkllnk) [reply app/models/spree/order.rb:175]: When I reviewed the pull request I also stumbled over `pre_discount_total` because the name seemed very general while we're only looking at vouchers at the moment. Being more specific can help people understand better what's happening. So, I would have named it like @dacook.

But I didn't raise this because a more general name can be more resilient to change later. I don't have the association of percentages and discount. I find the general concept of "pre discount" quite fitting.

So while I would have named it differently in the first place, I didn't think it was worth changing now. Happy either way but I personally wouldn't invest more work in this. :shrug: :cat2: :zzz:

Comment (rioug) [reply app/models/spree/order.rb:175]: I would keep as it is, even though "vouchers" are the only possible discount right now it still convey an understandable meaning .

Comment (dacook) [reply app/models/spree/order.rb:175]: Thanks for the discussion! Yes on further thought, "discounts" aren't necessarily a percentage. In my mind now, a "discount" is a price reduction provided by the store (whether flat or percentage). Whereas I imagined that a voucher could also be a gift from another person (eg I pay $10 for a gift voucher, for someone else to spend at the store). 

So I agree about not changing this now, we can discuss if/when we need to change it!

**Comment by drummer83** [comment]
> #### What should we test?
> 
>  - If an order has progressed to at least `payment` state then submitting a change to it's address should recalculate tax correctly if the change of address would change it's tax category. Note: which address gets used for tax purposes depends on an admin configuration setting.

Ok, someone needs to help me out here.
1. If the delivery address changes, then a different tax **rate** may apply. The tax **category** (of the product) remains unchanged. This is probably what is meant here, right?
2. Recalculating the tax due to a delivery address change is working before this PR already. What exactly has changed?
3. I am not aware of such admin configuration setting to select which address gets used. I know that tax rates and tax categories are set in admin configuration. The tax category of the product is selected on the edit page of the product. The tax rate is selected according to the delivery address. 

In summary I can't find anything that's not working - but I also didn't see any changes at all. So maybe we are good to merge this. Or please provide some more details for testing please, @Matt-Yorkley. Or do you have any insights, @filipefurtad0?

Thanks!

**Comment by rioug** [comment]
> * If the delivery address changes, then a different tax **rate** may apply. The tax **category** (of the product) remains unchanged. This is probably what is meant here, right?

Correct

> * Recalculating the tax due to a delivery address change is working before this PR already. What exactly has changed?

The code that takes care of that has been moved around, but it is expected to work as before. So If you didn't find anything broken that means we are all good. 
 
> * I am not aware of such admin configuration setting to select which address gets used. I know that tax rates and tax categories are set in admin configuration. The tax category of the product is selected on the edit page of the product. The tax rate is selected according to the delivery address.

I am not sure about this one.

**Comment by rioug** [comment]
@Matt-Yorkley can you clarify this ? see Konrad question above 
> Note: which address gets used for tax purposes depends on an admin configuration setting.

**Comment by Matt-Yorkley** [comment]
> The code that takes care of that has been moved around, but it is expected to work as before. So If you didn't find anything broken that means we are all good.

That's right :ok_hand: 

> Note: which address gets used for tax purposes depends on an admin configuration setting.

Hmmm that's strange, I thought there was an option to set the value of `Spree::Config[:tax_using_ship_address]` in the configuration page but can't find it now. Maybe it got removed at some point? 

Anyway, I think we can merge?
